### ISSUE TITLE
Args not defined for inbuilt methods

### ISSUE THREAD

**Issue by soummyaah** [pybee/voc#212]
For most of the methods, args have not been defined, which leads to the following problem.

5*6 is executed succesfully
However, 5.**mul**(6) throws an error. For every method, the args need to be defined.

The stack trace for:

```
# test.py
a = 4
print(a.__mul__(5))
```

Comes out to be:

```
Exception in thread "main" java.lang.IllegalArgumentException: wrong number of arguments
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.python.types.Function.invoke(Function.java:275)
    at org.python.types.Method.invoke(Method.java:44)
    at python.test.__init__.module$import(test.py:2)
    at python.test.__init__.main(test.py)
```

The operators are checked, however the underlying methods throw an error.
This arose in a conversion with @cflee 
He suggested to bridge this gap by modifying the tests/utils.py to test the math operators.

**Comment by freakboy3742** [comment]
This is fairly simple to do, it's just labour intensive - if you're interested in doing this, get in touch and I can give you some pointers.

**Comment by Melvie** [comment]
I'd like to take a crack at this bug ! Any advice/guidance would be awesome !

**Comment by Melvie** [comment]
@soummyaah Could you provide an advice on how to start solving this problem?

**Comment by eliasdorneles** [comment]
Hey @Melvie, I think I can help with this one.

So, the builtin functions are define in the file [python/common/org/Python.java](https://github.com/pybee/voc/blob/master/python/common/org/Python.java)

You can search that file for `NotImplementedError` to find the Python builtin functions that are missing implementation and find some that you'd like to implement. Basically, you'd have to write a Java version of the builtin function as it works on Python 3.4.

There is some documentation for how to write the Java functions with Python signatures here: https://voc.readthedocs.io/en/latest/internals/signatures.html

I highly recommend doing it in a test-driven manner, you can find the test files for the builtins under the [tests/builtins](https://github.com/pybee/voc/tree/master/tests/builtins) directory, you should add tests of your own for the function that you'll write.

**Comment by tulikaiam** [comment]
@freakboy3742 Can you help me out how to start off with it

**Comment by chandlerbaggett** [comment]
@freakboy3742 I'd be interested and have a ton of free time at the moment for labor-intensive coding. What pointers would you have? I have everything installed as well as having completed the tutorial.

Regards,
Chandler

**Comment by freakboy3742** [comment]
@chandlerbaggett Whenever a method is exposed to the Python namespace, you'll see an annotation on the method (see [this one](https://github.com/pybee/voc/blob/master/python/common/org/Python.java#L335) for abs()). This applies to every method in the builtins, plus all the methods on individual datatypes (like list()). These need to be updated to reflect the API and documentation that is exposed by Python. There's some summary documentation [here](https://voc.readthedocs.io/en/latest/internals/signatures.html) describing the options that are available for that annotation.

**Comment by chandlerbaggett** [comment]
@freakboy3742 Seems simple enough. I'll take a shot today and over the weekend to get all of the args defined. Thanks for explaining it for me.

Regards,
Chandler

**Comment by chandlerbaggett** [comment]
@freakboy3742 Or not an example, but which builtins might need it? I see where some of them throw an unimplemented error but I'm not exactly sure how to implement those that are listed.. I'd like to contribute so any tips/hints would be greatly appreciated.

Thanks!
Chandler

**Comment by eliasdorneles** [comment]
Hi @chandlerbaggett ! I believe all builtin methods in that Python.java file have already the proper annotations -- so they're good examples on how annotations are supposed to be used.

You'll find code that need updating in the classes implementing types here: https://github.com/pybee/voc/tree/master/python/common/org/python/types

Many methods have annotations with empty docstrings (search for `__doc__`).
Here is [one example in Dict.java](https://github.com/pybee/voc/blob/master/python/common/org/python/types/Dict.java#L363-L367).
It's missing the docstring, it should be the same as what you get if you do `dict.fromkeys.__doc__` in a Python 3 shell prompt.

Let me know if this isn't clear.

**Comment by chandlerbaggett** [comment]
@eliasdorneles @freakboy3742 All is clear! For clarification, though: sometimes the __doc__s will be surrounded by quotations and other times will be surrounded by apostrophes. Is it a big deal that we keep these consistent with the Python 3 shell prompt? I noticed that all of the __doc__ declarations in VOC's code have "" waiting for us so I've just been using quotations for all of them. Wanted to clarify before I got too far.
